src/provider/model/Model.ts

```diff
@@ -407,6 +420,77 @@ export class Model {
         }
     }
 
+    /**
+     * Attempts to find a given [[LiveApp]], or start an application if it cannot be found. Will test that the app
+     * fulfils given criteria, and returns either the [[LiveApp]] of the found or started app, or a value describing
+     * why our criteria was not fulfilled by the app (in which case no new app will be started)
+     *
+     * @param name The name of the application to use in any error messages
+     *
+     * @param identifyAppInfo Used to test if any currently running applications are the one we are looking for. Will
+     * be called with the `appInfo` of running [[LiveApp]]s. Will be called with an [[Application]], and should return
+     * true if this is the app we are looking for, false otherwise
+     *
+     * @param getAppInfo Will be called if we're not able to find the desired application among currently running
+     * applications. Should return the [[Application]] of an app we want to start
+     *
+     * @param testLiveApp In the case where the application we are looking for is already running (i.e.
+     * `identifyAppInfo` has found a match), this will be called to test if the [[LiveApp]] fulfils our desired
+     * criteria. This should return undefined if our criteria is fulfiled, otherwise T describing the failure. In the
+     * case where this is called, `getOrCreateLiveApp` will return either the [[LiveApp]] or the failure value
+     *
+     * @param testAppInfo In the case where the application we are looking for is not already running (i.e.
+     * `identifyAppInfo` has not found a match and we have an [[Application]] from `getAppInfo`), this will be called
+     * to test if the [[Application]] we want to start fulfils our desired criteria. This should return undefined if
+     * our criteria is fulfiled, otherwise T describing the failure. In the case where this is called,
+     * `getOrCreateLiveApp` will either start the [[Application]] and return the corresponding [[LiveApp]], or return
+     * the failure value
+     *
+     * @typeparam T Failure type, returned if one of ours tests for fails for the identified [[Application]]/[[LiveApp]]
+     */
```

Comment:
I think this is correct but I am surprised at the complexity of the api - essentially we have two callbacks that are then fed into two other callbacks, where those two other callbacks are only used once. I would perhaps naively imagined this could be simplified into a function that simply accepts two callbacks, and in the getOrCreateLiveAppByNameForIntent case, roll testLiveApp into identifyAppInfo and testAppInfo into getAppInfo. 

I suspect you have good reasons for this, though, so maybe I am wrong